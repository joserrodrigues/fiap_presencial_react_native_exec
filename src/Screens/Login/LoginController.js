import React from 'react';
import LoginView from './LoginView';
import {
    login,
} from '../../store/modules/login/actions';
import { useDispatch, useSelector } from 'react-redux';
import * as Yup from "yup";


const LoginController = () => {

    //Inicia o dispatch
    const dispatch = useDispatch();

    const isLoading = useSelector((state) => state.loginSaga.isLoading);
    const messageConnection = useSelector((state) => state.loginSaga.messageConnection);

    const signInSchema = Yup.object().shape({
        email: Yup.string().email("E-mail não válido").required("E-mail é obrigatório"),

        password: Yup.string()
            .required("Senha é obrigatório")
            .min(4, "Senha é curta - deveria ter ao menos 4 caracteres"),
    });

    submitForm = (values) => {
        dispatch(login(values.email,values.password));
    }
    
    return (
        <LoginView
            signInSchema={signInSchema}
            isLoading={isLoading}
            messageConnection={messageConnection}
            submitForm={submitForm} />
    );
};

export default LoginController;