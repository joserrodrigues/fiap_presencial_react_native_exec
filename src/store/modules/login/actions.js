/**
 * Objeto que reúne os Actions que alteram as informações no Reducer
 */

//Função que altera a Informação do Info
export function login(email, password) {
    return {
        type: 'login/Login',
        payload: { email, password },
    };
}
export function loginStart() {
    return {
        type: 'login/LoginStart'
    };
}
export function loginFail(status, message) {
    return {
        type: 'login/LoginFail',
        payload: { status, message },
    };
}
export function loginFinish(info) {
    return {
        type: 'login/LoginFinish',
        payload: { info },
    };
}

export function logout() {
    return {
        type: 'login/Logout',
    };
}
export function changeName(name) {
    return {
        type: 'login/ChangeName',
        payload: { name },
    };
}
export function stopLoadUserInfo(info) {
    return {
        type: 'login/StopLoadUserInfo',
        payload: { info },
    };
}


