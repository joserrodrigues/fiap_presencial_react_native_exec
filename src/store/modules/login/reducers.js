/**
 * Objeto que organiza o reducer do módulo busInfo
 */

//Importa a função produce do immer
import produce from 'immer';
import { useSetStorageItem } from '../../../Services/Storage/StorageServices';

//Seta o state inicial
const INITIAL_STATE = {
    isLoading: true,
    statusConnection: -1,
    messageConnection: "",
    userInfo: null
};

//Cria a função responsável por organizar o Reducer
export default function getInfo(state = INITIAL_STATE, { type, payload }) {

    //A função de uma maneira mais fácil o state com o draft
    // e retorna um novo state para o reducer
    return produce(state, (draft) => {
        switch (type) {

            case 'login/LoginStart': {      
                draft.isLoading = true;
                draft.statusConnection = 1;
                draft.messageConnection = "";
                break;
            }
            case 'login/LoginFail': {                
                draft.isLoading = false;
                draft.statusConnection = payload.status;
                draft.messageConnection = payload.message;
                break;
            }
            case 'login/LoginFinish': {
                draft.isLoading = false;
                draft.statusConnection = 0;                
                draft.userInfo = payload.info;
                draft.messageConnection = "";
                useSetStorageItem("userInfo",payload.info);         
                break;
            }
            case 'login/Logout': {
                draft.userInfo = null;  
                useSetStorageItem("userInfo", null);                 
                break;
            }
            case 'login/ChangeName': {
                draft.userInfo.name = payload.name;
                break;
            }
            case 'login/StopLoadUserInfo': {
                draft.isLoading = false;
                draft.statusConnection = 0;
                draft.userInfo = payload.info;
                break;
            }
            default:
        }
    });
}
