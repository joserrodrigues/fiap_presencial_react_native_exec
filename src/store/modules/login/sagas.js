/**
 * Objeto que reúne as sagas deste módulo
 */
//Importa os effects do Redux-Saga 
import { takeLatest, call, put, all } from 'redux-saga/effects';
import qs from 'qs';
import API from '../../../Services/APIs/Common/api';

// Importa as funções Actions que serão chamadas pelo Saga
import {
    loginStart,
    loginFail,
    loginFinish
} from './actions';

function* login({ payload }) {
    try {
        const { email, password } = payload;

        console.log(payload);
        yield put(loginStart());
    
        const data = qs.stringify({
            email: email,
            password: password
        });

        const dataReturn = yield call(API.post, 'storeProducts/login', data);
        
        console.log(dataReturn.status);
        console.log(dataReturn.data);

        if (dataReturn.status === 201){
            yield put(loginFail(-1, "Usuário ou login inválido"));
        } else if (dataReturn.status === 200){
            yield put(loginFinish(dataReturn.data));
        } else {
            yield put(loginFail(-98, "Erro desconhecido - Erro: " + dataReturn.status));    
        }        
    } catch (err) {
        console.log(err);
        yield put(loginFail(-99, "Erro ao realizar a conexão"));
    }
}

export default all([
    takeLatest('login/Login', login),
]);
