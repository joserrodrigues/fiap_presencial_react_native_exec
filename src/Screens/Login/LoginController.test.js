import React from 'react';
import renderer from 'react-test-renderer';
import api from '../../Services/APIs/Common/api';
import MockAdapter from "axios-mock-adapter";

import { render, fireEvent, waitFor } from '@testing-library/react-native';
import LoginController from './LoginController';
import store from '../../store/stores';
import { loginFinish } from '../../store/modules/login/actions';
import { Provider } from 'react-redux';


jest.setTimeout(30000);
jest.useFakeTimers();

const fakeLogin = {
    token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6InJ1YmVuc0BzY2',
    userId: '6207f78500e679092b819156',
    name: 'Rubens 2',
    phone: '11971745577'
};

beforeAll(() => {
    console.log(" IS MOCKING = " + global.__isMocking__);
    if (global.__isMocking__) {
        mock = new MockAdapter(api);
    }
});

afterEach(() => {
    if (global.__isMocking__) {
        mock.reset();
    }
});

describe('<LoginController />', () => {

    it('Check Initial Case', async () => {
        
        store.dispatch(loginFinish(null));

        const tree = renderer.create(
            <Provider store={store}>
                <LoginController />
            </Provider>
        ).toJSON();
        expect(tree.children.length).toBe(1);
        expect(tree).toMatchSnapshot();
    });

    it('Check Empty Form', async () => {

        await store.dispatch(loginFinish(null));

        const { getByText, getByTestId, getByPlaceholderText } = render(
            <Provider store={store}>
                <LoginController />
            </Provider>
        );

        const emailInput = getByPlaceholderText("E-mail");
        expect(emailInput).toBeTruthy();

        const passInput = getByPlaceholderText("Senha");
        expect(passInput).toBeTruthy();

        const button = getByTestId('loginButton');
        expect(button).toBeTruthy();
        fireEvent.press(button);

        let infoEmail = await waitFor(() => getByText('E-mail é obrigatório'));
        expect(infoEmail).toBeTruthy();
        let infoPass = await waitFor(() => getByText('Senha é obrigatório'));
        expect(infoPass).toBeTruthy();
    });

    it('Check Invalid E-mail', async () => {
        const { getByText, getByTestId, getByPlaceholderText } = render(
            <Provider store={store}>
                <LoginController />
            </Provider>
        );

        const emailInput = getByPlaceholderText("E-mail");
        expect(emailInput).toBeTruthy();

        const passInput = getByPlaceholderText("Senha");
        expect(passInput).toBeTruthy();

        const button = getByTestId('loginButton');
        expect(button).toBeTruthy();

        fireEvent.changeText(emailInput, 'hello');
        fireEvent.changeText(passInput, '123');
        fireEvent.press(button);

        await waitFor(() => {
            expect(getByText('E-mail não válido')).toBeTruthy();
            expect(getByText('Senha é curta - deveria ter ao menos 4 caracteres')).toBeTruthy();
        })
    });

    it('Check Invalid User', async () => {

        if (global.__isMocking__) {
            mock.onPost('/storeProducts/login').reply(201, fakeLogin);
        }

        const { getByText, getByTestId, getByPlaceholderText } = render(
            <Provider store={store}>
                <LoginController />
            </Provider>
        );

        const emailInput = getByPlaceholderText("E-mail");
        expect(emailInput).toBeTruthy();

        const passInput = getByPlaceholderText("Senha");
        expect(passInput).toBeTruthy();

        const button = getByTestId('loginButton');
        expect(button).toBeTruthy();

        fireEvent.changeText(emailInput, 'rubens@schoolguardian.app');
        fireEvent.changeText(passInput, '123456789');
        fireEvent.press(button);

        const element = await waitFor(() => getByText('Usuário ou login inválido'),
            { timeout: 1000500, interval: 500 });
        expect(element).toBeTruthy();
    });

    it('Check Valid User', async () => {

        if (global.__isMocking__) {
            mock.onPost('/storeProducts/login').reply(200, fakeLogin);
        }
        const { getByText, getByTestId, getByPlaceholderText } = render(
            <Provider store={store}>
                <LoginController />
            </Provider>
        );

        const emailInput = getByPlaceholderText("E-mail");
        expect(emailInput).toBeTruthy();

        const passInput = getByPlaceholderText("Senha");
        expect(passInput).toBeTruthy();

        const button = getByTestId('loginButton');
        expect(button).toBeTruthy();

        fireEvent.changeText(emailInput, 'rubens@schoolguardian.app');
        fireEvent.changeText(passInput, '123456');
        fireEvent.press(button);

        const element = await waitFor(() => getByText('Login com Sucesso'),
            { timeout: 1000500, interval: 500 });
        expect(element).toBeTruthy();
    });
});