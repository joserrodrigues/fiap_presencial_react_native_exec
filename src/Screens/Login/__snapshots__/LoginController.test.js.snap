// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginController /> Check Initial Case 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "flex-start",
    }
  }
>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Image
      resizeMode="cover"
      source={
        Object {
          "uri": "https://previews.123rf.com/images/chagin/chagin1501/chagin150100001/35151812-business-people-working-together.jpg",
        }
      }
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "height": undefined,
            "width": undefined,
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.6)",
          "flex": 1,
          "paddingBottom": 30,
          "paddingTop": 40,
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flex": 1,
            },
            Object {
              "paddingBottom": 0,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignContent": "center",
              "flex": 3,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "red",
                "fontSize": 40,
                "fontWeight": "bold",
                "margin": 20,
                "textAlign": "center",
              }
            }
          >
            RH App
          </Text>
        </View>
        <View
          style={
            Object {
              "flex": 4,
              "flexDirection": "column",
              "justifyContent": "center",
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgba(255,255,255,0.8)",
                "borderColor": "red",
                "borderRadius": 15,
                "borderWidth": 1,
                "margin": 35,
                "padding": 20,
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "paddingHorizontal": 10,
                    "width": "100%",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#86939e",
                      "fontSize": 16,
                      "fontWeight": "bold",
                    }
                  }
                >
                  E-mail
                </Text>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "borderBottomWidth": 1,
                      "borderColor": "#86939e",
                      "flexDirection": "row",
                      "transform": Array [
                        Object {
                          "translateX": 0,
                        },
                      ],
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "height": 40,
                        "justifyContent": "center",
                        "marginVertical": 4,
                        "paddingRight": 4,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "overflow": "hidden",
                        }
                      }
                    >
                      <View>
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "backgroundColor": "transparent",
                              "justifyContent": "center",
                            }
                          }
                        >
                          <Text />
                        </View>
                      </View>
                    </View>
                  </View>
                  <TextInput
                    allowFontScaling={true}
                    editable={true}
                    id={1}
                    onChangeText={[Function]}
                    onSubmitEditing={[Function]}
                    placeholder="E-mail"
                    placeholderTextColor="#86939e"
                    rejectResponderTermination={true}
                    secureTextEntry={false}
                    style={
                      Object {
                        "color": "#242424",
                        "flex": 1,
                        "fontSize": 18,
                        "minHeight": 40,
                      }
                    }
                    testID="RNE__Input__text-input"
                    underlineColorAndroid="transparent"
                  />
                </View>
                <Text
                  style={
                    Object {
                      "color": "#ff190c",
                      "fontSize": 12,
                      "margin": 5,
                    }
                  }
                />
              </View>
              <View
                style={
                  Object {
                    "paddingHorizontal": 10,
                    "width": "100%",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#86939e",
                      "fontSize": 16,
                      "fontWeight": "bold",
                    }
                  }
                >
                  Senha
                </Text>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "borderBottomWidth": 1,
                      "borderColor": "#86939e",
                      "flexDirection": "row",
                      "transform": Array [
                        Object {
                          "translateX": 0,
                        },
                      ],
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "height": 40,
                        "justifyContent": "center",
                        "marginVertical": 4,
                        "paddingRight": 4,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "overflow": "hidden",
                        }
                      }
                    >
                      <View>
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "backgroundColor": "transparent",
                              "justifyContent": "center",
                            }
                          }
                        >
                          <Text />
                        </View>
                      </View>
                    </View>
                  </View>
                  <TextInput
                    allowFontScaling={true}
                    editable={true}
                    id={1}
                    onChangeText={[Function]}
                    onSubmitEditing={[Function]}
                    placeholder="Senha"
                    placeholderTextColor="#86939e"
                    rejectResponderTermination={true}
                    secureTextEntry={true}
                    style={
                      Object {
                        "color": "#242424",
                        "flex": 1,
                        "fontSize": 18,
                        "minHeight": 40,
                      }
                    }
                    testID="RNE__Input__text-input"
                    underlineColorAndroid="transparent"
                  />
                </View>
                <Text
                  style={
                    Object {
                      "color": "#ff190c",
                      "fontSize": 12,
                      "margin": 5,
                    }
                  }
                />
              </View>
              <View
                style={
                  Object {
                    "marginBottom": 20,
                  }
                }
              />
            </View>
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "height": 60,
                  "justifyContent": "space-evenly",
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "borderRadius": 3,
                      "overflow": "hidden",
                    },
                    Object {
                      "borderRadius": 3,
                    },
                    Object {
                      "flex": 1,
                      "marginHorizontal": 20,
                      "marginVertical": 10,
                    },
                    false,
                  ]
                }
              >
                <View
                  accessibilityRole="button"
                  accessibilityState={
                    Object {
                      "busy": false,
                      "disabled": false,
                    }
                  }
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "opacity": 1,
                    }
                  }
                  testID="loginButton"
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "red",
                        "borderColor": "#2089dc",
                        "borderRadius": 10,
                        "borderWidth": 0,
                        "flexDirection": "row",
                        "justifyContent": "center",
                        "padding": 8,
                      }
                    }
                  >
                    <Text
                      style={
                        Object {
                          "color": "white",
                          "fontSize": 18,
                          "paddingVertical": 1,
                          "textAlign": "center",
                        }
                      }
                    >
                      Login
                    </Text>
                  </View>
                </View>
              </View>
            </View>
            <View>
              <Text
                style={
                  Object {
                    "color": "#fa5757",
                    "fontSize": 13,
                    "fontWeight": "bold",
                    "textAlign": "center",
                  }
                }
              >
                Login com Sucesso
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
